lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -53,32 +54,41 @@ namespace Org.Apache.REEF.IMRU.OnREEF.Driver
     /// <typeparam name="TMapInput">Map Input</typeparam>
     /// <typeparam name="TMapOutput">Map output</typeparam>
     /// <typeparam name="TResult">Result</typeparam>
-    internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult> : IObserver<IDriverStarted>,
-        IObserver<IAllocatedEvaluator>, IObserver<IActiveContext>, IObserver<ICompletedTask>, IObserver<IFailedEvaluator>
+    /// <typeparam name="TDataHandler">Data Handler type in IInputPartition</typeparam>
+    internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TDataHandler> 
+        : IObserver<IDriverStarted>,
+        IObserver<IAllocatedEvaluator>,
+        IObserver<IActiveContext>,
+        IObserver<ICompletedTask>,
+        IObserver<IFailedEvaluator>,
+        IObserver<IFailedContext>,
+        IObserver<IFailedTask>,
+        IObserver<IRunningTask>
     {
-        private static readonly Logger Logger = Logger.GetLogger(typeof(IMRUDriver<TMapInput, TMapOutput, TResult>));
+        private static readonly Logger Logger =
+            Logger.GetLogger(typeof(IMRUDriver<TMapInput, TMapOutput, TResult, TDataHandler>));
 
         private readonly ConfigurationManager _configurationManager;
-        private readonly IPartitionedInputDataSet _dataSet;
+        private readonly int _totalMappers;
         private readonly IEvaluatorRequestor _evaluatorRequestor;
         private ICommunicationGroupDriver _commGroup;
         private readonly IGroupCommDriver _groupCommDriver;
         private readonly TaskStarter _groupCommTaskStarter;
-        private readonly ConcurrentStack<string> _taskIdStack;
         private readonly ConcurrentStack<IConfiguration> _perMapperConfiguration;
-        private readonly Stack<IPartitionDescriptor> _partitionDescriptorStack;
         private readonly int _coresPerMapper;
         private readonly int _coresForUpdateTask;
         private readonly int _memoryPerMapper;
         private readonly int _memoryForUpdateTask;
         private readonly ISet<IPerMapperConfigGenerator> _perMapperConfigs;
-        private readonly ConcurrentBag<ICompletedTask> _completedTasks;
+        private readonly IList<ICompletedTask> _completedTasks = new List<ICompletedTask>();
         private readonly int _allowedFailedEvaluators;
         private int _currentFailedEvaluators = 0;
-        private bool _reachedUpdateTaskActiveContext = false;
         private readonly bool _invokeGC;
+        private bool _areImruTasksCompleted = false;
+        private int _numberOfReadyTasks = 0;
+        private readonly object _lock = new object();
```

Comment:
If this lock is only for _completedTasks, you might want to use more specific name like _completedTasksLock